harness/determined/core/_checkpoint.py

```diff
@@ -283,11 +304,34 @@ def _upload_sharded(
         # Decide if our rank is the lowest rank trying to upload this ckpt_dir.
         want_upload = file_uid and all_file_uids.index(file_uid) == self._dist.rank
 
-        # Decide what we are going to upload.
+        # Decide what we are going to upload; get all files in the ckpt_dir to the
+        # uploading rank and filter them based on the selectors set for local ranks.
+        # For detailed description see download().
         if want_upload:
             assert ckpt_dir
             resources = self._storage_manager._list_directory(ckpt_dir)
+
+            def _selector(path: str) -> bool:
+                if not selector:
+                    return True
+
+                _ = self._dist.broadcast_local(path)
```

Comment:
this appears to assume that the uploading rank is a local chief, which is not a safe assumption